jmespath/visitor.py

```diff
@@ -1,13 +1,20 @@
 import operator
 
 from jmespath import functions
+from jmespath.functions import is_listlike
 from jmespath.compat import string_type
 from numbers import Number
 
 
+def is_ndarray(arg):
+    return hasattr(arg, "__array_finalize__")
+
+
 def _equals(x, y):
     if _is_special_number_case(x, y):
         return False
+    elif is_ndarray(x) or is_ndarray(y):
```

Comment:
what happens if one of them isn't a ndarray (e.g., a string or a number that is of a mismatching dtype), does numpy still return a boolean array?

Reply:
numpy comparison if either are ndarray, equal if both types are trivially convertable

```python
>>> a1 = np.array([1,2,3], dtype='int32')
>>> a1 == [1,2,3]
array([ True,  True,  True])
>>> a1 == ["1", "2", "3"]
array([False, False, False])
>>> ["1", "2", "3"] == a1
array([False, False, False])
>>> a1 == np.array([1,2,3], dtype='int64')
array([ True,  True,  True])
```

Reply:
Thanks, that's useful! I had more in mind comparisons against scalars, which also seem to work **most** of the time, but not all:

```
>>> a1 = np.array((1, 2, 3), dtype='int32')
>>> a1 == 3.
array([False, False,  True])
>>> a1 == 'bah'
<stdin>:1: FutureWarning: elementwise comparison failed; returning scalar instead, but in the future will perform elementwise comparison
False
```

That last one will break :(

Reply:
On my version of numpy this last one already does this elementwise. I've updated to instead of assuming a return type use dynamic handling:

```python
def _arraylike_all(arg):
    return arg.__array__().all() if hasattr(arg, "__array__") else arg
    
def _equals(x, y):
    if _is_special_number_case(x, y):
        return False
    elif is_array(x) or is_array(y):
        return _arraylike_all(x == y)
    else:
        return x == y
```